kotlinx-coroutines-core/common/src/sync/ThroughputLimiter.kt

```diff
@@ -0,0 +1,368 @@
+    public suspend fun tryAcquire(timeout: Duration): Boolean
+
+    /**
+     * Get counts of permit resolutions, granted, denied and such.
+     * @see ThroughputCounterEventType
+     */
+    public fun stats(): Map<ThroughputCounterEventType, Long>
+    public fun resetStats()
+}
+
+/**
+ * Limit throughput of events, per interval, to be at most equal to the argument eventsPerInterval.
+ * When the limit is passed, calls are suspended until the calculated point in time when it's
+ * okay to pass the rate limiter.
+ */
+@SinceKotlin("1.5")
+@OptIn(ExperimentalTime::class)
+public interface IntervalLimiter : ThroughputLimiter
+
+@SinceKotlin("1.5")
+@OptIn(ExperimentalTime::class)
+public fun intervalLimiter(eventsPerInterval: Int, interval: Duration, warmupPeriod: Duration? = null): IntervalLimiter =
+    IntervalLimiterImpl(eventsPerInterval = eventsPerInterval, interval = interval, warmupPeriod = warmupPeriod)
+
+@SinceKotlin("1.5")
+@OptIn(ExperimentalTime::class)
+internal class IntervalLimiterImpl(
+    eventsPerInterval: Int,
+    interval: Duration = Duration.seconds(1),
+    private val timeSource: NanoTimeSource = NanoTimeSourceImpl,
+    private val delay: suspend (Long) -> Unit = ::delay,
+    warmupPeriod:Duration? = null
+) : IntervalLimiter {
+
+    init {
+        require(interval.inWholeMilliseconds > 5) {
+            "Interval has to be at least 5 ms. The overhead of having locks and such in place if enough to render this moot."
+        }
+        require(interval.inWholeDays <= 1) {
+            "Interval has to be less than 1 day"
+        }
+        require(interval.inWholeNanoseconds / eventsPerInterval > 1) {
+            "Interval segment is not allowed to be less than one"
+        }
+        require(warmupPeriod?.let { interval <= warmupPeriod } != false){
+            "Interval has to be greater or equal to the warmup period"
+        }
+    }
+
+    private var warmupMark:NanoTimeMark? = warmupPeriod?.let { timeSource.markNow() + it }
+    //private val oneNano = Duration.nanoseconds(1)
+    private val mutex = Mutex()
+    private val _interval = Duration.nanoseconds(interval.inWholeNanoseconds)
+    private val eventSegment = _interval.div(eventsPerInterval)
+    private val counter = ThroughputCounter(CoroutineScope(Dispatchers.Default))
+
+    // Mutable state, access through mutex
+    private var cursor: NanoTimeMark = timeSource.markNow()
+    private var intervalStartCursor: NanoTimeMark = timeSource.markNow() - eventSegment
+    private var intervalEndCursor: NanoTimeMark = intervalStartCursor + _interval
+    // End Mutable state
+
+    override suspend fun acquire(): Long = acquire(permits = 1)
+    override suspend fun acquire(permits: Int): Long {
+        if (permits < 0) throw IllegalArgumentException("You need to ask for at least zero permits")
+
+        val now: NanoTimeMark = timeSource.markNow()
+        val permitDuration = if (permits == 1) eventSegment else eventSegment.times(permits)
```

Comment:
Good question!
It seems from some testing that using some extreme numbers this breaks down completely!

Reply:
I think ill need to drop the actual time based cursors in favour of count based + timeMarker, or something.. Ill have to teat this out a bit and see what works and what will have to fail validation. Like grabbing 10k permits where each permit would be equal to one year, might be out of more that one validation range..

Reply:
I guess that perhaps it's best to deny any request where the number of permits requested is larger than the number of events per unit of time that is allowed?

Reply:
In the documentation of Guava - it at least implies that big permit asks is not a bad thing, it will however block other upcoming requests..
The main issue that I have at the moment is that kotlin time does not seem to be usable for handling the nano time roll over..
I'm able to restrict the user from creating MASSIVE permits that overflow the Duration.. But I need to be able to handle the situation when the nano time moves naturaly (or the cursors move) from Long.MAX_VALUE over to Long.MIN_VALUE, and also be able to keep track of the offset between now and the cursor, in case one has moved into the next iteration and the other has not.. Supporting an offset of one should be the same as supporting an offset of n.

At this point I'm not writing any code.. Just meditating on the problem.
I do this for fun, when I have time to spare. I wont rush it, or else it won't be fun ;-)

Reply:
I hear you!

Reply:
I think I was able to solve the time roll-over issues